warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java

```diff
@@ -81,96 +95,168 @@ public Object visit(ASTOrNode node, Object data) {
+                    // if we have an hdfs cache directory and if past the fst threshold, then create the fst and replace the list with an assignment
+                    if (rangeNodesByField.isEmpty() && fstHdfsUri != null && (eqNodes.size() >= config.getMaxOrExpansionFstThreshold())) {
+                        URI fstPath = createFst(values);
+                        markers.add(new ExceededOrThresholdMarkerJexlNode(field, fstPath));
+                        eqNodes = null;
+                    } else if (eqNodes.size() >= config.getMaxOrExpansionThreshold()) {
+                        markers.add(new ExceededOrThresholdMarkerJexlNode(field, values, null));
+                        eqNodes = null;
                     }
                     
-                    marker = new ExceededOrThresholdMarkerJexlNode(field, fstPath);
-                } else {
-                    marker = new ExceededOrThresholdMarkerJexlNode(field, values);
+                    // handle range nodes separately
+                    if (rangeNodes.size() >= config.getMaxOrRangeThreshold()) {
+                        Collection<Range> ranges = new TreeSet<>();
+                        rangeNodes.forEach(rangeNode -> ranges.add(rangeNodeToRange(rangeNode)));
+                        
+                        markers.add(new ExceededOrThresholdMarkerJexlNode(field, null, ranges));
+                        rangeNodes = null;
+                    }
+                } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | IOException e) {
+                    QueryException qe = new QueryException(DatawaveErrorCode.LARGE_FIELDED_LIST_ERROR, e);
+                    throw new DatawaveFatalQueryException(qe);
+                }
+                
+                // add in any unused eq nodes
+                if (eqNodes != null) {
+                    for (JexlNode child : eqNodes) {
+                        JexlNode copiedChild = (JexlNode) child.jjtAccept(this, data);
+                        if (copiedChild != null) {
+                            children.add(copiedChild);
+                        }
+                    }
                 }
-                children.add(marker);
+                
+                // add in any unused range nodes
+                if (rangeNodes != null) {
+                    for (JexlNode child : rangeNodes) {
+                        JexlNode copiedChild = (JexlNode) child.jjtAccept(this, data);
+                        if (copiedChild != null) {
+                            children.add(copiedChild);
+                        }
+                    }
+                }
+                
+                children.addAll(markers);
             }
             // else simply add the subset back into the children list
             else {
-                children.addAll(subsetChildrenCopies);
+                // recurse on the eq children in this subset
+                for (JexlNode child : eqNodes) {
```

Comment:
This pattern appears here a few times where we loop over a collection of jexl nodes, apply this visitor, and add the child to the children list if the result is not null. Making a method like `collectChildren(children, eqNodes);` could reduce the duplicated logic in a few places.

Reply:
I agree.  Updated.